Automattic/wp-calypso
# DevDocs / State: Add DevDocs screen for searching state selectors

Related: #5954

This pull request seeks to add a new screen to the [DevDocs section](https://wpcalypso.wordpress.com/devdocs) to enable a developer to search for available selectors. Search is fuzzy, enabling generic searches like "getSite" to return the `getSelectedSite` selector, but in less precedence than the `getSite` selector.

Search View | Single View
---|---
![Search](https://cloud.githubusercontent.com/assets/1779930/20353862/29f87256-abea-11e6-9861-a795d72c97cc.png)|![Detail](https://cloud.githubusercontent.com/assets/1779930/20353866/2d213bb6-abea-11e6-8f0e-822881cb8c80.png)

__Implementation Notes:__

Fuzzy searching uses [Fuse.js](http://fusejs.io). The DevDocs section already uses another similar library [Lunr.js](http://lunrjs.com) for full text search, though this library does not have great support for fuzzy searching, and it was important for me to support.

The selector parsing is cached, though not to the filesystem as in the case of document searching. Any request to any section of DevDocs will start "priming" the selector cache and this cache will be used for the remainder of the server process.

Included are several examples of selectors that will be removed prior to merging. See #5954.

__Testing Instructions:__

1. Navigate to [DevDocs State Selectors](http://calypso.localhost:3000/devdocs/selectors)
2. Note that all selectors from the `client/state/selectors` are visible with description
3. Enter a search term
4. Note that fuzzy results are shown in order of matching precedence
5. Click a result
6. Note that selector is shown with arguments are return type
7. Click Back
8. Note that you're returned to search results

__Open Questions / Ideas:__

- If we were to parse the file using a proper JavaScript parser, we could interpret arguments even for selectors without good JSDoc, though I argue it's better to encourage good documentation by using them as a prerequisite for arguments displaying
- Search only applies to selector name and description, though we could also apply this to parameters or return types (e.g. search all selectors that accept `siteId` as a parameter)
- What other [JSDoc tags](http://usejsdoc.org) would be useful to include aside from `@param` and `@return`? Another good option may be `@see` for external resources.

cc @withinboredom
